sympy/solvers/solvers.py

```diff
@@ -1389,15 +1389,38 @@ def _solve(f, *symbols, **flags):
     # captured here (for reference below) in case flag value changes
     flags['check'] = checkdens = check = flags.pop('check', True)
 
+    if f.is_Add:
```

Comment:
When this was `if not f.is_Mul:` there were some odd test failures. And if you don't watch out for hollow factoring, like `f(2+4*x)->f(2*(1+2*x))`, you can get into recursion trouble.